android/src/main/java/com/swmansion/reanimated/nodes/Node.java

```diff
@@ -76,54 +76,92 @@ public void removeChild(Node child) {
     }
   }
 
-  protected void markUpdated() {
+  public void onDrop() {
+    // no-op
+  }
+
+  public @Nullable List<Node> getChildrenInContext(EvalContext context) {
+    return mChildren;
+  }
+
+  public EvalContext switchContextWhileUpdatingIfNeeded(EvalContext context, Node lastVisited) {
+    return context;
+  }
+
+  protected void markUpdated(EvalContext context) {
     UiThreadUtil.assertOnUiThread();
-    mUpdateContext.updatedNodes.add(this);
+    context.updatedNodes.add(this);
     mNodesManager.postRunUpdatesAfterAnimation();
   }
 
-  protected final void dangerouslyRescheduleEvaluate() {
-    mLastLoopID = -1;
-    markUpdated();
+  protected final void dangerouslyRescheduleEvaluate(EvalContext context) {
+    context.lastLoopsIDs.put(mNodeID, (long) -1);
+    markUpdated(context);
   }
 
-  protected final void forceUpdateMemoizedValue(Object value) {
-    mMemoizedValue = value;
-    markUpdated();
+  protected final void forceUpdateMemoizedValue(Object value, EvalContext context) {
+    context.memoizedValues.put(mNodeID, value);
+    markUpdated(context);
   }
 
-  private static void findAndUpdateNodes(Node node, Set<Node> visitedNodes, Stack<FinalNode> finalNodes) {
+  private static void findAndUpdateNodes(Node node, Set<Node> visitedNodes, Stack<FinalNode> finalNodes, Stack<EvalContext> contexts, Node lastVisited) {
     if (visitedNodes.contains(node)) {
       return;
-    } else {
-      visitedNodes.add(node);
+    }
+    visitedNodes.add(node);
+
+    EvalContext currentContext = contexts.peek();
+    List<Node> children = node.getChildrenInContext(currentContext);
+    EvalContext newContext = node.switchContextWhileUpdatingIfNeeded(currentContext, lastVisited);
```

Comment:
I'd rename it to sth more like `contextForUpdatingChildren` as it switches nothing and only returns a context that should be used for updating children. The switch is done here by adding that context to the stack. Lets keep method name more descriptive.